NSB injection in MC R1 waveform

gabemery opened the pull request:
An option to inject NSB pulses in existing Monte-Carlo simulation at the level of calibrated waveforms (R1) is implemented.
It :

  - extracts the simulation NSB from the analysed file
  - finds the missing NSB compared to the target data file comparing the pedestal variance (or take it from the config)
  - randomly assigns a number of pulses to inject in each waveform using Poisson statistic
  - randomly draws an exact injection time (uniform) and normalisation (using the spe normalisation pdf)
  - modifies the waveform accordingly accounting for AC coupling by lowering the baseline uniformly

It is controlled by adding the following block in the configuration file:
"waveform_nsb_tuning": {
"nsb_tuning":true, #false to turn it off, is ignored for real data
"nsb_tuning_ratio": 0.89, #Additional fraction of the original NSB to be added, if not given will be extracted from the target_data file
"target_data": "", # File from which the NSB correction should be extracted. To avoid running it for each MC file, using the nsb_tuning_ratio input with pre-extracted value will often be preferred
"spe_location":"", # File containing the single photo-electron normalisation pdf
"pulse_template_location": "" # file containing the spe pulse template in the form "time pulseLG pulseHG" or "time pulse"
}

The following shows a couple of results obtained with prod5_trans 80 MC taking a gamma diffuse and a proton simtel file (run1). The target data here is the run 2969, subrun 0000 of the Crab Nebula observation of November 20th 2020. They are produced using another branch from the likelihood method development where it was first developed. Assuming no error during the copy of the changes to the current branch it should be equivalent but dedicated tests of this version will be needed. The NSB increase is by ~72%.

Current injection looks as below on a couple of waveforms (without and with signal, event_id 11003 of prod5_trans80 gamma_20deg_180deg_run1___cta-prod5-lapalma_4LSTs_MAGIC_desert-2158m_mono_cone6.simtel.gz) :
![image](https://user-images.githubusercontent.com/35919817/146374131-e6fd61a7-e39a-463f-a3de-893256aa1ece.png)
![image](https://user-images.githubusercontent.com/35919817/146374155-abd39f00-4321-4517-80d1-275cec63da70.png)


The title contains the pixel number and number of injected pulses in a time window starting 20 samples before the waveform (relative to the 0 of the pulse template which is ~10 ns before its maximum).

The effect on integrated charge is shown below (on the next event of the same file, id 16705)
![image](https://user-images.githubusercontent.com/35919817/146374081-939eb22f-111d-4296-8ab3-70f93224810e.png)

A first check of the validity of the correction was done by comparing the distribution of all extracted charges of all pixels in the low intensity range. The comparison below shows : a non modified gamma diffuse MC file, the same file after tuning to our target data set, a tuned proton file, and the target dataset. The NSB correction was independently extracted for both MC file during the processing.

![image](https://user-images.githubusercontent.com/35919817/146374181-83550e3d-30d8-474d-aa4d-6a6644f82165.png)

I look at the low intensity pixels since they are dominated by pure background pixels. Higher intensities contain signal so spectral normalisation as a function of the particle type would be needed. It is clearly visible here that the correction improves the data/MC agreement in the charge extracted for low intensity pixels.

This is a redo of #819 with a branch from the base repository instead of a fork to allow the download of the test dataset .

gabemery commented:
Original PR comments :
--- from mojalejo : 
Hi 'at'gabemery, thanks for this very useful tool!

I did not check the code, but from the description of the procedure, I think it does everything it has to do.

How is the test done? You mention gamma and proton MC, but you should compare to noise-only MC (e.g. plotting only pixels which have no p.e.'s from Cherenkov light, if that info is available), so it should not matter if it is gamma or proton.

In the test you show the resulting noise seems marginally higher, the difference being surely unimportant, but just out of curiosity: can it be further tuned? (I mean manually - I understand the automatic approach may have this kind of small differences, and they are irrelevant)

I think it would be good to add the example configuration for Crab (disabled by default) to the standard json file.
Then, as for the availability of the reference files for the pulse shape and gain fluctuations, shall we add them together with the code, at least those we use in the coming major MC production?

--- from gabemery :
Hi 'at'moralejo , In the test, the code is run on either the gamma or proton file without pre-computed correction to the NSB. It is instead determined using a new function 'calculate_required_additional_nsb' (which is largely a copy of your function calculate_noise_parameters, I plan to try to reduce duplicated code here if possible). Then, taking the output dl1 files, I load the charge in each pixels (in image) for all events and create an histogram. Here two things are done : first I limit to 10 p.e. to look mostly at pure noise pixels, second I normalise the distributions to account for the different number of events and noise pixels in each files.

As you said, looking only at pure noise pixels would remove the need to look only at low intensity. It could probably be done by applying a cleaning on data and MC (mimicking the use of only signal less pixels in MC with data will be harder, but possible).

I am not sure I see this high resulting noise you mention. After tuning, the ratio of data/MC distributions of low intensity pixels are visually the same up to ~6 p.e. Then there is a bit more data but this can be signal. Either removing those or using a realistic renormalisation of the MC could solve this discrepancy.
It is possible to manually tune quite freely by using the "nsb_tuning_ratio" parameter in the config file (its presence deactivate the automatic procedure). The extraction method can be called in external scripts to get the needed correction. I actually think it would be the best workflow in the end to extract the needed correction first (averaging it over the full target dataset instead of a single file, and maybe also averaging over multiple MC files). Then give it in the config instead of re-extracting it 11000 times.

Yes I think the reference pulse and gain fluctuation files should be included. I saw that the pulse is already in lstchain/data/oversampled_pulse_LST_8dynode_pix6_20200204.dat but without the time information. From the commit it seems it will be removed in 0.8.0 so it could be replaced by the version which works with my code.

moralejo commented:
> 
> As you said, looking only at pure noise pixels would remove the need to look only at low intensity. It could probably be done by applying a cleaning on data and MC (mimicking the use of only signal less pixels in MC with data will be harder, but possible).

But in MC (well, depending on the sim_telarray settings) you know which pixels have  p.e.'s from Cherenkov light, so you can obtain pure-noise pixel spectra. 
Then, in data, you have interleaved pedestal events, which again are pure noise. So I do not see any need for the cleaning approach you suggest.

> 
> I am not sure I see this high resulting noise you mention. After tuning, the ratio of data/MC distributions of low intensity pixels are visually the same up to ~6 p.e. Then there is a bit more data but this can be signal. 

Yes, indeed, if you are not applying a cut to remove those, it can be signal.

> 
> Yes I think the reference pulse and gain fluctuation files should be included. I saw that the pulse is already in lstchain/data/oversampled_pulse_LST_8dynode_pix6_20200204.dat but without the time information. From the commit it seems it will be removed in 0.8.0 so it could be replaced by the version which works with my code.

Ok, seems good to me. If that ever changes (not likely in the near future) we will have to introduce "epochs" in the correction, based on runnumber or whatever.

maxnoe commented:
The reference pulse shape is available from the `SubarrayInformation` provided by the `LSTEventSource` via `SubarrayInfo -> TelescopeDescription -> CameraDescription -> CameraReadout`:

```
In [6]: from ctapipe_io_lst import LSTEventSource

In [7]: lst1_subarray = LSTEventSource.create_subarray(4)

In [8]: lst1_subarray.tel[1].camera.readout
Out[8]: CameraReadout(camera_name='LSTCam', sampling_rate='1.024 1 / ns', reference_pulse_sample_width=0.02441406 ns)

In [9]: lst1_subarray.tel[1].camera.readout.reference_pulse_shape
Out[9]: 
array([[-6.234646e-05, -7.277727e-05, -8.958578e-05, ...,  0.000000e+00,
         0.000000e+00,  0.000000e+00],
       [-3.890991e-03, -4.093170e-03, -4.226685e-03, ...,  0.000000e+00,
         0.000000e+00,  0.000000e+00]])
```

gabemery commented:
I have added the pulse template extraction from event source (thanks @maxnoe), the SPE file in lstchain/data and solved a couple of conflicts with the new lstchain version. After that I performed a new test of the method using the branch proposed for merging ( previous plots were from an independent branch in 0.7.3 where I first implemented this code).
Looking at the distribution of charge in signal less pixels only following @moralejo  advice ( 0 pe pixels in simu and flat field events in data) I saw a divergence between the data and corrected MC (71% injection). I thus reverted an empirical correction I had found in the previous comparison : the NSB correction is set to be proportional to the difference of pedestal variance. In the case of subrun 2969.0000 it is 52%

In a previous comparison I had found that using the square root of this value improved the agreement but it was likely related to the comparison being done with signal (or might be an effect of going from 0.7.3 to 0.8.2post1).

![NSB_tuning_charge_per_pixels_01_20_revertedcorrection_v082post1_withratio](https://user-images.githubusercontent.com/35919817/148522430-7088abe3-b74c-4e8b-b263-b05f523aa6f6.png)
